src/Components/Forms/src/FieldIdentifier.cs

```diff
@@ -100,42 +100,69 @@ private static void ParseAccessor<T>(Expression<Func<T>> accessor, out object mo
+                // so, given that it embeds captured values such as "this". We could consider special-casing
+                // for "() => something.Member" and building a cache keyed by "something.GetType()" with values
+                // of type Func<object, object> so we can cheaply map from "something" to "something.Member".
+                var modelLambda = Expression.Lambda(memberExpression.Expression);
+                var modelLambdaCompiled = (Func<object?>)modelLambda.Compile();
+                var result = modelLambdaCompiled();
+                if (result is null)
+                {
+                    throw new ArgumentException("The provided expression must evaluate to a non-null value.");
+                }
+                model = result;
+            }
+            else
             {
                 throw new ArgumentException("The provided expression must evaluate to a non-null value.");
             }
-            model = constantExpression.Value;
         }
-        else if (memberExpression.Expression != null)
+        else if (accessorBody is MethodCallExpression methodCallExpression
+            && methodCallExpression.Method.IsSpecialName
+            && methodCallExpression.Method.Name == "get_Item")
         {
-            // It would be great to cache this somehow, but it's unclear there's a reasonable way to do
-            // so, given that it embeds captured values such as "this". We could consider special-casing
-            // for "() => something.Member" and building a cache keyed by "something.GetType()" with values
-            // of type Func<object, object> so we can cheaply map from "something" to "something.Member".
-            var modelLambda = Expression.Lambda(memberExpression.Expression);
-            var modelLambdaCompiled = (Func<object?>)modelLambda.Compile();
-            var result = modelLambdaCompiled();
-            if (result is null)
+            // Field name will be the index value, enclosed in brackets
+            fieldName = $"[{methodCallExpression.Arguments[0]}]";
+
+            // Get a reference to the model object
+            // i.e., given a value like "(something)[1]", determine the runtime value of "(something)",
+            if (methodCallExpression.Object != null)
+            {
+                var modelLambda = Expression.Lambda(methodCallExpression.Object);
+                var modelLambdaCompiled = (Func<object?>)modelLambda.Compile();
+                var result = modelLambdaCompiled();
+                if (result is null)
+                {
+                    throw new ArgumentException("The provided expression must evaluate to a non-null value.");
+                }
+                model = result;
+            }
+            else
             {
                 throw new ArgumentException("The provided expression must evaluate to a non-null value.");
             }
```

Comment:
This block of code looks identical to lines 118-136. Could the duplication be factored out into a method?